File: backend/domains/agent/presentation/schemas/listing_studio.py
"""Listing Studio API 请求/响应 Schema。"""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field


class RunStepBody(BaseModel):
    capability_id: str = Field(..., description="能力 ID")
    user_input: dict[str, Any] = Field(default_factory=dict)
    model_id: str | None = Field(None, description="用户模型 UUID 或系统模型 ID")
    meta_prompt: str | None = None
    prompt_template_id: str | None = None


class OptimizePromptBody(BaseModel):
    capability_id: str = Field(..., description="能力 ID")
    user_input: dict[str, Any] = Field(default_factory=dict)
    meta_prompt: str | None = None
    model_id: str | None = Field(None, description="用户模型 UUID 或系统模型 ID")


class RunPipelineBody(BaseModel):
    inputs: dict[str, Any] = Field(default_factory=dict)
    steps: list[str] | None = None
    session_id: str | None = None
    model_overrides: dict[str, str] | None = Field(
        default=None,
        description="按 capability_id 指定各步 model_id",
    )


class CreateTemplateBody(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    content: str | None = None
    prompts: list[str] | None = None


class UpdateTemplateBody(BaseModel):
    name: str | None = Field(None, min_length=1, max_length=100)
    content: str | None = None
    prompts: list[str] | None = None


class CreateImageGenTaskBody(BaseModel):
    prompts: list[dict[str, Any]] = Field(default_factory=list)
    job_id: str | None = None
    model_id: str | None = None
    provider: str | None = None
    size: str | None = None
    reference_image_url: str | None = None
    strength: float | None = None


class ListingStudioJobStepResponse(BaseModel):
    id: str
    job_id: str
    sort_order: int
    capability_id: str
    input_snapshot: dict[str, Any] | None = None
    output_snapshot: dict[str, Any] | None = None
    meta_prompt: str | None = None
    generated_prompt: str | None = None
    prompt_used: str | None = None
    prompt_template_id: str | None = None
    status: str
    error_message: str | None = None
    created_at: str | None = None
    updated_at: str | None = None


class ListingStudioJobResponse(BaseModel):
    id: str
    user_id: str | None = None
    session_id: str | None = None
    title: str | None = None
    status: str
    created_at: str | None = None
    updated_at: str | None = None
    steps: list[ListingStudioJobStepResponse] = Field(default_factory=list)


class ListingStudioJobListResponse(BaseModel):
    items: list[ListingStudioJobResponse]
    total: int
    skip: int
    limit: int


class ListingStudioCapabilitiesResponse(BaseModel):
    capabilities: list[dict[str, Any]]
    execution_layers: list[list[str]]


class RunPipelineResponse(BaseModel):
    job_id: str
    status: str
    message: str
    poll_url: str


class OptimizePromptResponse(BaseModel):
    capability_id: str
    optimized_prompt: str


class UploadImageResponse(BaseModel):
    url: str
    content_type: str
    size_bytes: int


def job_response(data: dict[str, Any]) -> ListingStudioJobResponse:
    return ListingStudioJobResponse.model_validate(data)


def job_list_response(
    items: list[dict[str, Any]],
    *,
    total: int,
    skip: int,
    limit: int,
) -> ListingStudioJobListResponse:
    return ListingStudioJobListResponse(
        items=[job_response(item) for item in items],
        total=total,
        skip=skip,
        limit=limit,
    )


__all__ = [
    "CreateImageGenTaskBody",
    "CreateTemplateBody",
    "ListingStudioCapabilitiesResponse",
    "ListingStudioJobListResponse",
    "ListingStudioJobResponse",
    "ListingStudioJobStepResponse",
    "OptimizePromptBody",
    "OptimizePromptResponse",
    "RunPipelineBody",
    "RunPipelineResponse",
    "RunStepBody",
    "UpdateTemplateBody",
    "job_list_response",
    "job_response",
]

File: frontend/src/types/listing-studio.ts
/**
 * Listing Studio Types - Listing 创作工作流类型
 */

export type ListingStudioJobStatus = 'draft' | 'running' | 'completed' | 'failed' | 'partial'

export type ListingStudioStepStatus =
  | 'pending'
  | 'prompt_generating'
  | 'prompt_ready'
  | 'running'
  | 'completed'
  | 'failed'

export interface ListingStudioJobStep {
  id: string
  job_id: string
  sort_order: number
  capability_id: string
  input_snapshot: Record<string, unknown> | null
  output_snapshot: Record<string, unknown> | null
  meta_prompt: string | null
  generated_prompt: string | null
  prompt_used: string | null
  prompt_template_id: string | null
  status: ListingStudioStepStatus
  error_message: string | null
  created_at: string | null
  updated_at: string | null
}

export interface ListingStudioJob {
  id: string
  user_id: string | null
  session_id: string | null
  title: string | null
  status: ListingStudioJobStatus
  created_at: string | null
  updated_at: string | null
  steps?: ListingStudioJobStep[]
}

export interface ListingStudioJobListResponse {
  items: ListingStudioJob[]
  total: number
  skip: number
  limit: number
}

export interface ListingStudioCapability {
  id: string
  name: string
  sort_order?: number
  model_type?: 'text' | 'image'
  output_key?: string
  dependencies?: string[]
  input_fields?: string[]
  meta_prompt_params?: { key: string; label: string }[]
  required_features?: string[]
}

export interface ListingStudioCapabilitiesResponse {
  capabilities: ListingStudioCapability[]
  execution_layers: string[][]
}

export interface ListingStudioPromptTemplate {
  id: string
  user_id: string | null
  capability_id: string
  name: string
  content: string | null
  prompts: string[] | null
  created_at: string | null
  updated_at: string | null
}

export interface RunStepBody {
  capability_id: string
  user_input: Record<string, unknown>
  model_id?: string | null
  meta_prompt?: string | null
  generated_prompt?: string | null
  prompt_template_id?: string | null
  phase?: 'generate_prompt' | 'execute' | 'full'
}

export interface OptimizePromptBody {
  capability_id: string
  user_input: Record<string, unknown>
  meta_prompt?: string | null
  model_id?: string | null
}

export interface OptimizePromptResponse {
  capability_id: string
  optimized_prompt: string
}

export interface RunPipelineBody {
  inputs?: Record<string, unknown>
  steps?: string[] | null
  session_id?: string | null
  model_overrides?: Record<string, string> | null
}

export interface RunPipelineResponse {
  job_id: string
  status: string
  message: string
  poll_url: string
}

export interface ProductImageGenTask {
  id: string
  job_id?: string | null
  status: string
  prompts: Record<string, unknown>[]
  result_images: { slot: number; url: string; error?: string; skipped?: boolean }[] | null
  error_message: string | null
  created_at: string | null
  updated_at?: string | null
}

export interface ImageGenProvider {
  id: string
  name: string
  default_size: string
  sizes: string[]
  supports_reference_image: boolean
}

/** Listing 创作页用户输入（草稿 persist） */
export interface ListingStudioInputs {
  product_link?: string
  competitor_link?: string
  product_name?: string
  keywords?: string
  image_urls?: string[]
}

export interface ProductImageGenTaskListResponse {
  items: ProductImageGenTask[]
  total: number
  skip: number
  limit: number
}
